[ADD] Form to create manual share lines

remytms opened the pull request:
This add the ability to create share lines that are not releated to a subscription request and that does not appear in subscription register.

The goal is to let user add shares that he get's in the past from the same structure.

remytms committed:
[ADD] iwp_base: Creation mode on share lines

remytms committed:
[ADD] iwp_website: Creation mode in share wallet

remytms commented:
Not sure if the term 'auto' and 'manual' are good names. If not clear, suggestion are welcome.

houssine78 commented:
maybe use platform instead of auto and private instead of manual ?

remytms committed:
[ADD] iwp_website: Form to create manual share line

remytms commented:
@robinkeunen This is my proposition for formulars in Odoo. Can be improved of course, but I think this is a strong base. :)

remytms committed:
[ADD] iwp_website: Action to delete manual share line

robinkeunen commented on iwp_website/controllers/form.py:
```diff
@@ -0,0 +1,145 @@
+        """Return True if the form has no errors, or False otherwise."""
+        return self.is_bound and not self.errors
+
+    def full_clean(self):
```
function name should start with a verb.

remytms replied:
Ici j'ai repris la même nomenclature que Django. Tu aurais préféré `clean_full()` ou `clean_all()` ? Mais dans l'idée cette fonction pourrait même être privée. On ne l'appelle pas depuis l'extérieur.

robinkeunen replied:
Yes j'aurais bien vu `clean_all` ou `clean_all_fields`

robinkeunen commented on iwp_website/controllers/form.py:
```diff
@@ -0,0 +1,145 @@
+        self._clean_fields()
+        self._clean_form()
+
+    def add_error(self, name, err):
```
private?

remytms replied:
Oui, je l'aurais bien vu en privé. Mais chez Django elle ne l'est pas, je me suis dit qu'il y avait une raison…

robinkeunen commented on iwp_website/controllers/form.py:
```diff
@@ -0,0 +1,145 @@
+        try:
+            cleaned_data = self.clean()
+        except FormValidationError as err:
+            self.add_error("__all__", err)
```
Je comprends pas pourquoi tu ajoutes cette erreur sous la clé `__all__`

remytms replied:
Les erreurs de validation générer par la méthode `clean()` (elle ne fait rien par défaut) sont des erreurs qui ne sont pas liées à un champ particulier mais à un ensemble de champs.

Par exemple, si deux champs ne peuvent pas être remplis en même temps, c'est dans la méthode clean qu'on va faire le check, car le validateur d'un des deux champs ne travaille que sur la valeur du champ et pas les valeurs des autres champs.

Ces erreurs concernent donc le formulaire dans son ensemble et sont donc ajouté à la clé `__all__` (tous les champs).

robinkeunen replied:
👍

robinkeunen reviewed (approved):
Ca a l'air vraiment pas mal mec. On essaiera de généraliser et de l'appliquer sur médor!

remytms committed:
[ADD] iwp_website: Button to take manual share

remytms committed:
[FIX] iwp_website: date issue

The method `fromisoformat` does not exist on python 3.5, only since
python 3.7.
